packages/flutter_tools/lib/src/test/flutter_tester_device.dart

```diff
@@ -0,0 +1,337 @@
+  }
+
+  @override
+  Future<void> kill() async {
+    globals.printTrace('test $id: Terminating flutter_tester process');
+    _process?.kill(io.ProcessSignal.sigkill);
+
+    globals.printTrace('test $id: Shutting down test harness socket server');
+    await _server?.close(force: true);
+    await finished;
+  }
+
+  @override
+  Future<void> get finished async {
+    final int exitCode = await _exitCode.future;
+
+    // On Windows, the [exitCode] and the terminating signal have no correlation.
+    if (globals.platform.isWindows) {
+      return;
+    }
+
+    // ProcessSignal.SIGKILL. Negative because signals are returned as negative
+    // exit codes.
+    if (exitCode == -9) {
+      // We expect SIGKILL (9) because we could have tried to [kill] it.
+      return;
+    }
+    throw TestDeviceException(_getExitCodeMessage(exitCode), StackTrace.current);
+  }
+
+  Uri get _ddsServiceUri {
+    return Uri(
+      scheme: 'http',
+      host: (host.type == InternetAddressType.IPv6 ?
+        InternetAddress.loopbackIPv6 :
+        InternetAddress.loopbackIPv4
+      ).host,
+      port: debuggingOptions.hostVmServicePort ?? 0,
+    );
+  }
+
+  @visibleForTesting
+  @protected
+  Future<DartDevelopmentService> startDds(Uri uri) {
+    return DartDevelopmentService.startDartDevelopmentService(
+      uri,
+      serviceUri: _ddsServiceUri,
+      enableAuthCodes: !debuggingOptions.disableServiceAuthCodes,
+      ipv6: host.type == InternetAddressType.IPv6,
+    );
+  }
+
+  /// Binds an [HttpServer] serving from `host` on `port`.
+  ///
+  /// Only intended to be overridden in tests.
+  @protected
+  @visibleForTesting
+  Future<HttpServer> bind(InternetAddress host, int port) => HttpServer.bind(host, port);
+
+  @protected
+  @visibleForTesting
+  Future<StreamChannel<String>> get remoteChannel async {
+    assert(_server != null);
+
+    try {
+      final HttpRequest firstRequest = await _server.first;
+      final WebSocket webSocket = await WebSocketTransformer.upgrade(firstRequest);
+      return _webSocketToStreamChannel(webSocket);
+    } on Exception catch (error, stackTrace) {
+      throw TestDeviceException('Unable to connect to flutter_tester process: $error', stackTrace);
+    }
+  }
+
+  @override
+  String toString() {
+    final String status = _process != null
+        ? 'pid: ${_process.pid}, ${_exitCode.isCompleted ? 'exited' : 'running'}'
+        : 'not started';
+    return 'Flutter Tester ($status) for test $id';
+  }
+}
+
+void _pipeStandardStreamsToConsole({
+  @required int id,
+  @required Process process,
+  @required Future<void> reportObservatoryUri(Uri uri),
+}) {
+  const String observatoryString = 'Observatory listening on ';
+  for (final Stream<List<int>> stream in <Stream<List<int>>>[
+    process.stderr,
+    process.stdout,
+  ]) {
+    stream
+        .transform<String>(utf8.decoder)
+        .transform<String>(const LineSplitter())
+        .listen(
+      (String line) async {
```

Comment:
Is it possible to re-use the existing ObservatoryDiscovery class here?

Reply:
It's possible but the pieces don't quite fit together, not sure if it's worth doing now. `ProtocolDiscovery.observatory` requires a `DeviceLogReader`, but we don't quite have that here. I'd say doing that would be easier when we eventually (?) refactor to use the `FlutterTesterDevice` abstraction here, where pieces like the log reader are already created for us.